lib/experimental/class-wp-webfonts.php

```diff
@@ -14,238 +14,332 @@
-	 *                                 or a font-family as a string.
-	 * @return string|false The font slug on success, or false if the font-family cannot be determined.
+	 * @param string $font_family_handle The font family's handle for this variation.
+	 * @param string $variation_handle   Variation's handle.
+	 * @return bool True when registered to the given font family. Else false.
 	 */
-	public static function get_font_slug( $to_convert ) {
-		if ( is_array( $to_convert ) ) {
-			if ( isset( $to_convert['font-family'] ) ) {
-				$to_convert = $to_convert['font-family'];
-			} elseif ( isset( $to_convert['fontFamily'] ) ) {
-				$to_convert = $to_convert['fontFamily'];
-			} else {
-				_doing_it_wrong( __METHOD__, __( 'Could not determine the font family name.', 'gutenberg' ), '6.0.0' );
-				return false;
-			}
+	private function is_variation_registered( $font_family_handle, $variation_handle ) {
+		if ( ! isset( $this->registered[ $font_family_handle ] ) ) {
+			return false;
 		}
 
-		return sanitize_title( $to_convert );
+		return in_array( $variation_handle, $this->registered[ $font_family_handle ]->deps, true );
 	}
 
 	/**
-	 * Validate a webfont.
+	 * Adds a variation as a dependency to the given font family.
 	 *
-	 * @since 6.0.0
+	 * @since X.X.X
 	 *
-	 * @param array $webfont The webfont arguments.
+	 * @param string $font_family_handle The font family's handle for this variation.
+	 * @param string $variation_handle   The variation's handle.
+	 */
+	private function add_dependency( $font_family_handle, $variation_handle ) {
+		$this->registered[ $font_family_handle ]->deps[] = $variation_handle;
+	}
+
+	/**
+	 * Validates and sanitizes a variation.
+	 *
+	 * @since X.X.X
 	 *
-	 * @return array|false The validated webfont arguments, or false if the webfont is invalid.
+	 * @param string $font_family_handle The font family's handle for this variation.
+	 * @param array  $variation          An array of variation properties to add.
+	 * @return false|array Validated variation on success. Else, false.
 	 */
-	public function validate_webfont( $webfont ) {
-		$webfont = wp_parse_args(
-			$webfont,
-			array(
-				'provider'     => 'local',
-				'font-family'  => '',
-				'font-style'   => 'normal',
-				'font-weight'  => '400',
-				'font-display' => 'fallback',
-			)
+	private function validate_variation( $font_family_handle, $variation ) {
+		$defaults = array(
+			'provider'     => 'local',
+			'font-family'  => $font_family_handle,
+			'font-style'   => 'normal',
+			'font-weight'  => '400',
+			'font-display' => 'fallback',
 		);
 
-		// Check the font-family.
-		if ( empty( $webfont['font-family'] ) || ! is_string( $webfont['font-family'] ) ) {
-			trigger_error( __( 'Webfont font family must be a non-empty string.', 'gutenberg' ) );
-			return false;
-		}
+		$defaults = apply_filters( 'wp_webfont_variation_defaults', $defaults );
```

Comment:
If we add a filter here, we should probably document it as well... Though tbh, I think the filter would be more useful just a few lines below and affect the `$variation` instead of the `defaults`? 🤔

Reply:
I'm thinking more about this filter. It's invoked for every single variation that gets validated. But as these are default properties, filtering over and over again is wasteful of server resources. Instead, these defaults could be set in the constructor, i.e. run once.

Reply:
Looks a lot better now! 👍